hazelcast-client/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheInvalidationTest.java

```diff
@@ -0,0 +1,336 @@
+import com.hazelcast.test.HazelcastParametersRunnerFactory;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.SlowTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import javax.cache.spi.CachingProvider;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test that near cache invalidation events are delivered when a cache is:
+ * <ul>
+ *     <li><code>clear</code>ed</li>
+ *     <li><code>destroy</code>ed (with <code>Cache.destroy()</code></li>
+ *     <li>destroyed by its client-side <code>CacheManager.destroyCache</code></li>
+ * </ul>
+ * and <b>not delivered</b> when a cache is closed (<code>Cache.close()</code>).
+ * Respective operations are tested when executed either from member or client-side Cache proxies with the exception of
+ * CacheManager.destroyCache, in which case the near-cache is already destroyed on the client-side and the listener
+ * registration is removed <b>before</b> the invocation for destroying the cache is sent to the member, so no event can be
+ * received.
+ */
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
+@Category({SlowTest.class})
+public class ClientNearCacheInvalidationTest extends HazelcastTestSupport {
+
+    static final String DEFAULT_CACHE_NAME = "com.hazelcast.client.cache.nearcache.ClientNearCacheInvalidationTest";
+
+    // time to wait until invalidation event is delivered (when used with assertTrueEventually)
+    // and time to wait when testing that no invalidation event is delivered (used with assertTrueAllTheTime)
+    static final int TIMEOUT = 5;
+
+    // when true, invoke operations which are supposed to deliver invalidation events from a Cache instance on a member
+    // otherwise use the client-side proxy.
+    @Parameter
+    public boolean invokeCacheOperationsFromMember;
+
+    @Parameter(1)
+    public InMemoryFormat inMemoryFormat;
+
+    @Parameters(name = "fromMember:{0}, format:{1}")
+    public static Collection<Object[]> parameters() {
+        return Arrays.asList(new Object[][]{
+                {false, InMemoryFormat.BINARY},
+                {false, InMemoryFormat.OBJECT},
+                {true, InMemoryFormat.BINARY},
+                {true, InMemoryFormat.OBJECT},
+        });
+    }
+
+    private HazelcastInstance member;
+    private TestHazelcastFactory hazelcastFactory;
+    private NearCacheTestContext testContext;
+
+    @Before
+    public void setup() {
+        hazelcastFactory = new TestHazelcastFactory();
+        member = hazelcastFactory.newHazelcastInstance(getConfig());
```

Comment:
start more than one members and you will see when_cacheDestroyed_invalidationEventIsReceived test fails.